### DISCUSSION TITLE
Integrating Langium with a UI Editor like React-flow

### DISCUSSION THREAD

**Discussion by Lemour-sudo** [eclipse-langium/langium#1373] [Q&A] [answered]
I am using Langium to define a DSL that will consist of operators. As an example operator, I have an addition operator that will be expected to take in two numbers `Add: 'Add' '(' value1=INT ',' value2=INT ')';`.

Now I want to extract this operator's definition from the Lanigum-generated files and use it in a UI editor (React-flow as an example) that will represent this operator as a node in the UI.

The UI will require the operator's keyword, inputs and their types, and the return type. I do see I can get the keywords and input arguments from both the 'ast.ts' and grammar files that Langium generates. But I cannot find a way to specify what return type should be expected from the operator. Is there a way to tell Langium what type of input a rule returns and have Langium add this return type to the rule's metadata in the generated grammar file?

I cannot define a return type for the the addition operator because Langium does not seem to support return types for normal parser rules. Another option I was considering is if there is a way to add additional metadata for each parser rule such that it is added to the data generated in the grammar file.

**Comment by msujew** [comment]
Hey @Lemour-sudo,

first off - Langium is not a modelling framework and trying to use it as such isn't exactly recommended. 

That being said, the type information for a given grammar is derived in [here](https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/grammar/type-system/ast-collector.ts). You can use the code yourself to derive additional information about your own grammar types at runtime if needed.

> I cannot define a return type for the the addition operator because Langium does not seem to support return types for normal parser rules. Another option I was considering is if there is a way to add additional metadata for each parser rule such that it is added to the data generated in the grammar file.

You should be able to adjust the return type by [using declared types](https://langium.org/docs/sematic-model/#declared-types). Alternatively you can use the [`infers` syntax](https://langium.org/docs/sematic-model/#parser-rules).

**Reply by Lemour-sudo** [reply]
Thank you for the reply.

I tried the declared types and they do not seem ideal in my case as they set the parameters required for an operator's tag to be successfully parsed. The `infers` syntax just references a parser rule to another type in the grammar.

Can you elaborate more on what you mean by Langium not being a modeling framework and how that is related to outputting data types that can be used to create visual nodes in a UI editor?

**Reply by msujew** [reply]
> I tried the declared types and they do not seem ideal in my case as they set the parameters required for an operator's tag to be successfully parsed. The infers syntax just references a parser rule to another type in the grammar.

Well, what do you need in your case? I'm not really sure what you mean by setting a parser rules _return type_. By definition, a parser rule returns an AST node of the shape that is specified in the parser rule itself (i.e. through assignments).

> Can you elaborate more on what you mean by Langium not being a modeling framework and how that is related to outputting data types that can be used to create visual nodes in a UI editor?

Langium, first and foremost, is a _language engineering tool_. It is supposed to be used as a tool to build language servers or tooling that is very similar in its design to language servers; I.e. usually things with a workflow that look like this:
1. Parsing from a textual presentation
2. Perform semantic work based on the AST (stuff like _go to reference_, or identifying whether the text content is semantically valid, stuff like type errors, etc.)
3. (optional) generate code or build an interpreter/debugger based on the AST.

With modeling framework I'm talking very specifically about Langium's predecessor Xtext and it's fairly heavy inclusion of the [Eclipse Modeling Framework](https://eclipse.dev/modeling/emf/). What you seem to look for is a way to use Langium models as storage/representation of your UI models. This is specifically out of scope for Langium.

**Reply by Lemour-sudo** [reply]
Thanks for the explanation on what the main purposes of Langium are.

What I meant by _return type_ was in a programming sense. In the addition example, I can tell that the parser should look for parameters of numbers. Thinking about in the sense of a function as in programming, where the function is expected to take in arguments of certain types and return a value of a certain type - can we do something similar when defining a parser rule that a particular type is expected to be returned when executing the rule's action?

I may be thinking about the parser rules in the wrong way. Do let me know what you think.

**Reply by msujew** [reply, marked as answer]
> What I meant by return type was in a programming sense. In the addition example, I can tell that the parser should look for parameters of numbers. 

You would need to write a type system yourself. Type systems effectively sit on top of a parsed AST and provide additional information about types of the parsed AST elements. I would recommend you to take a look at our [Lox example language repo](https://github.com/TypeFox/langium-lox) and especially the [type system related code](https://github.com/TypeFox/langium-lox/tree/main/src/language-server/type-system).

The lox language itself is from the [crafting interpreters](https://craftinginterpreters.com/) book. I can only recommend to take a look at that if you're completely new to anything compiler related.

> I may be thinking about the parser rules in the wrong way. Do let me know what you think.

Probably - you mix up syntax and semantics. A parser works purely on a syntactical level, while something like a type system is semantically driven. Everything related to the semantics of a language builds on top of its syntax (represented by the AST - the **A**bstract **S**yntax **T**ree), but a parser only ever works on syntax, not semantics.
